watchtower/lookout/interface.go

```diff
@@ -0,0 +1,77 @@
+package lookout
+
+import (
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightningnetwork/lnd/chainntnfs"
+	"github.com/lightningnetwork/lnd/watchtower/wtdb"
+)
+
+// Service abstracts the lookout functionality, supporting the ability to start
+// and stop. All communication and actions are driven via the database or chain
+// events.
+type Service interface {
+	// Start safely starts up the Interface.
+	Start() error
+
+	// Stop safely stops the Interface.
+	Stop() error
+}
+
+// BlockFetcher supports the ability to fetch blocks.
+type BlockFetcher interface {
+	// FetchBlockByHash fetchs the block given the target block hash.
+	GetBlock(*chainhash.Hash) (*wire.MsgBlock, error)
+}
+
+// DB abstracts the required persistent calls expected by the lookout. DB
+// provides the ability to search for state updates that correspond to breach
+// transactions confirmed in a particular block.
+type DB interface {
+	// GetLastMatchedEpoch returns the last block epoch at which the tower
+	// performed a match. If no match has been done, a nil epoch will be
+	// returned.
+	GetLastMatchedEpoch() (*chainntnfs.BlockEpoch, error)
+
+	// FindMatches searches its database for any state updates matching the
+	// provided breach hints. To be correct, a breach hint should be
+	// provided for each transaction included at the provided block epoch.
+	// If any matches are found, they will be returned along with encrypted
+	// blobs so that justice can be exacted.
+	FindMatches(*chainntnfs.BlockEpoch, []wtdb.BreachHint) ([]wtdb.Match, error)
+}
+
+// EpochRegistrar supports the ability to register for events corresponding to
+// newly created blocks.
+type EpochRegistrar interface {
+	// RegisterBlockEpochNtfn registers for a new block epoch subscription.
+	// The implementation must support historical dispatch, starting from
+	// the provided chainntnfs.BlockEpoch when it is non-nil. The
+	// notifications should be delivered in-order, and deliver reorged
+	// blocks.
+	RegisterBlockEpochNtfn(
+		*chainntnfs.BlockEpoch) (*chainntnfs.BlockEpochEvent, error)
+}
+
+// EpochSource delivers an in-order stream of blocks as they are seen on the
+// network.
+type EpochSource interface {
+	// Epochs provides an in-order stream of reorg-safe blocks.
+	Epochs() <-chan *Epoch
+
+	// Start safely spins up the EpochSource, starting from the provided
+	// epoch if it is non-nil.
+	StartFromEpoch(*chainntnfs.BlockEpoch) error
+
+	// Stop safely shuts down the EpochSource.
+	Stop() error
+}
+
+// Punisher handles the construction and publication of justice transactions
+// once they have been detected by the Service.
+type Punisher interface {
+	// Punish accepts a JusticeDescriptor, constructs the justice
+	// transaction, and publishes the transaction to the network so it can
+	// be mined.
+	Punish(*JusticeDescriptor, <-chan struct{}) error
```

Comment:
reliable publication on startup isn't implemented at this point, but that is the plan. yes the second parameter is the quit channel, so that we can shutdown the punisher if it's in the middle of watching for spends (when that is implemented)